Remove DSL deprecated methods and others

artembilan opened the pull request:
Since the DSL code is new in this version there is no reason to keep deprecated method

* Remove deprecated `resequence()` and `aggregate()` in the `IntegrationFlowDefinition`
* Remove `InternalAggregatingMessageHandler` with the `MessageGroupProcessorWrapper` delegation logic in favor of newly introduced `AbstractCorrelatingMessageHandler.setOutputProcessor()`
* Rework `Promise` Gateway Docs to the `Mono`
* Rename `AsyncGatewayTests` "promise" words to "mono"

artembilan committed:
Remove DSL deprecated methods and others

Since the DSL code is new in this version there is no reason to keep deprecated method

* Remove deprecated `resequence()` and `aggregate()` in the `IntegrationFlowDefinition`
* Remove `InternalAggregatingMessageHandler` with the `MessageGroupProcessorWrapper` delegation logic in favor of newly introduced `AbstractCorrelatingMessageHandler.setOutputProcessor()`
* Rework `Promise` Gateway Docs to the `Mono`
* Rename `AsyncGatewayTests` "promise" words to "mono"
* Resolve `com.rabbitmq.client.FlowListener` deprecation

artembilan commented:
Do not merge yet.
I have something for the `RecipientListRouter` as well.

artembilan committed:
RecipientListRouter DSL refactoring

Previously there was a DSL specific `DslRecipientListRouter` to overcome the lack of setters in the target `RecipientListRouter` and its `Recipient`

* Add `channelName` variant for the `Recipient`
* Add several new `addRecipient()` methods to the `RecipientListRouter`

All those improvements allow to avoid extra bridge (or adapter) component between Java DSL and target `RecipientListRouter`

* Make `LambdaMessageProcessor` and `GatewayMessageHandler` as `public` classes and move them to the appropriate packages

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/gateway/GatewayMessageHandler.java:
```diff
@@ -31,24 +29,24 @@
  * @since 5.0
  */
-class GatewayMessageHandler extends AbstractReplyProducingMessageHandler implements Lifecycle {
+public class GatewayMessageHandler extends AbstractReplyProducingMessageHandler implements Lifecycle {
```
Needs Javadocs now it's public.

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/handler/LambdaMessageProcessor.java:
```diff
@@ -40,7 +39,7 @@
  * @since 5.0
  */
-class LambdaMessageProcessor implements MessageProcessor<Object>, BeanFactoryAware {
+public class LambdaMessageProcessor implements MessageProcessor<Object>, BeanFactoryAware {
```
Needs Javadocs

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/router/RecipientListRouter.java:
```diff
@@ -93,12 +93,15 @@ public void setChannels(List<MessageChannel> channels) {
+		Assert.notEmpty(recipients, "'recipients' must not be empty");
 		ConcurrentLinkedQueue<Recipient> originalRecipients = this.recipients;
 		this.recipients.clear();
 		this.recipients.addAll(recipients);
```
This should be atomic - looks like it was broken by INT-2856; although not 

There is a small timing hole where there's no recipients, or an incomplete list.

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/router/RecipientListRouter.java:
```diff
@@ -93,12 +93,15 @@ public void setChannels(List<MessageChannel> channels) {
 		this.recipients.addAll(recipients);
+		if (getBeanFactory() != null) {
+			this.recipients.forEach(recipient -> recipient.setChannelResolver(getChannelResolver()));
+		}
```
This needs to be done before the atomic update above.

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/dsl/RecipientListRouterSpec.java:
```diff
@@ -89,8 +97,14 @@ public RecipientListRouterSpec recipientMessageSelector(String channelName, Mess
+			messageSelector = (MessageSelector) selector;
+		}
+		else {
+			messageSelector = new MethodInvokingSelector(new LambdaMessageProcessor(selector, null));
```
Is this strictly correct? Don't we have to check it's a Lambda first? Someone might implement his own `GenericSelector<P>` that is not a `MessageSelector`.

artembilan replied:
It might be called `Lamba` by accident, but the purpose is still there single method invocation purpose for provided `payloadType`. My idea to rely on the  strict `GenericSelector` implementation and bypass `MessagingMethodInvokerHelper` which is based on the SpEL and a bit slower than direct method invocation.

garyrussell replied:
I realize the `LMP` will work for any functional interface. My point is, this won't work...

	.recipient("foo", new Foo())

	public static class Foo implements GenericSelector<Message<?>> {

		@Override
		public boolean accept(Message<?> source) {
			return false;
		}

		public boolean foo(String bar) {
			return false;
		}

	}

...because the LMP will select the second method not the first.

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/dsl/AggregatorSpec.java:
```diff
@@ -57,17 +52,16 @@ public AggregatorSpec processor(Object target) {
 	 */
 	public AggregatorSpec processor(Object target, String methodName) {
 		super.processor(target);
 		return this.outputProcessor(methodName != null
```
Remove `this.`; also line 75.

garyrussell reviewed (changes requested):
Some minor issues.

artembilan committed:
Address PR comments

artembilan committed:
Add `isLambda()` condition to `RecipientListRouterSpec` to avoid ambiguity for the provided `GenericSelector` impl

artembilan committed:
Fix CheckStyle violations and one more `isLambda()`` in the `RecipientListRouterSpec`

Add `Assert` for method ambiguity to the `LambdaMessageProcessor` to be sure that we get deal only with the Function Interface impl

garyrussell commented:
Merged as 092d876fae368ff34a1af79a58bf869e3b5de2a3 with checkstyle fixes.
